source/game/kart/CollisionGroup.cc

```diff
@@ -0,0 +1,205 @@
+    m_worldPos = m_relPos + pos;
+}
+
+void Hitbox::reset() {
+    m_worldPos.setZero();
+    m_lastPos.setZero();
+    m_relPos.setZero();
+}
+
+void Hitbox::setRadius(f32 radius) {
+    m_radius = radius;
+}
+
+void Hitbox::setBspHitbox(const BSP::Hitbox *hitbox) {
+    m_bspHitbox = hitbox;
+}
+
+void Hitbox::setWorldPos(const EGG::Vector3f &pos) {
+    m_worldPos = pos;
+}
+
+void Hitbox::setLastPos(const EGG::Vector3f &pos) {
+    m_lastPos = pos;
+}
+
+void Hitbox::setLastPos(const EGG::Vector3f &scale, const EGG::Matrix34f &pose) {
+    f32 yScaleFactor = scale.y;
+    EGG::Vector3f scaledPos = m_bspHitbox->position;
+    scaledPos.x *= scale.x;
+    scaledPos.z *= scale.z;
+
+    if (scale.y != scale.z && scale.y < 1.0f) {
+        scaledPos.y += (1.0f - scale.y) * m_radius;
+        yScaleFactor = scale.z;
+    }
+
+    scaledPos.y *= yScaleFactor;
+    m_lastPos = pose.ps_multVector(scaledPos);
+}
+
+const BSP::Hitbox *Hitbox::bspHitbox() const {
+    return m_bspHitbox;
+}
+
+EGG::Vector3f Hitbox::worldPos() const {
+    return m_worldPos;
+}
+
+EGG::Vector3f Hitbox::lastPos() const {
+    return m_lastPos;
+}
+
+f32 Hitbox::radius() const {
+    return m_radius;
+}
+
+CollisionGroup::CollisionGroup() : m_bspHitboxCount(0), m_hitboxScale(1.0f) {
+    m_collisionData.reset();
+}
+
+CollisionGroup::~CollisionGroup() {
+    delete[] m_hitboxes->data();
+    delete m_hitboxes;
+}
+
+f32 CollisionGroup::initHitboxes(const std::array<BSP::Hitbox, 16> &hitboxes) {
+    u16 bspHitboxCount = 0;
+
+    for (const auto &hitbox : hitboxes) {
+        if (parse<u16>(hitbox.enable)) {
+            ++bspHitboxCount;
+        }
+    }
+
+    m_bspHitboxCount = bspHitboxCount;
+
+    m_hitboxes = new std::span<Hitbox>(new Hitbox[m_bspHitboxCount], m_bspHitboxCount);
+
+    for (u16 hitboxIdx = 0; hitboxIdx < m_hitboxes->size(); ++hitboxIdx) {
+        (*m_hitboxes)[hitboxIdx].setBspHitbox(&hitboxes[hitboxIdx]);
+    }
+
+    return computeCollisionLimits();
+}
+
+f32 CollisionGroup::computeCollisionLimits() {
+    EGG::Vector3f max;
+
+    for (const auto &hitbox : *m_hitboxes) {
+        const BSP::Hitbox *bspHitbox = hitbox.bspHitbox();
+
+        if (bspHitbox->enable == 0) {
+            continue;
+        }
+
+        max = max.maximize(bspHitbox->position.abs() + bspHitbox->radius);
+    }
+
+    // Get largest component of the vector
+    f32 maxComponent = max.z;
+
+    if (max.x <= max.y) {
+        if (max.z < max.y) {
+            maxComponent = max.y;
+        }
+    } else if (max.z < max.x) {
+        maxComponent = max.x;
+    }
+
+    m_boundingRadius = maxComponent;
+
+    return max.z * 0.5f;
+}
+
+void CollisionGroup::createSingleHitbox(f32 radius, const EGG::Vector3f &relPos) {
+    m_bspHitboxCount = 1;
+    m_hitboxes = new std::span<Hitbox>(new Hitbox[m_bspHitboxCount], m_bspHitboxCount);
+
+    // TODO: Do we need for loop if this is just one?
+    // And how exactly will we identify to free the BSP::Hitbox on destruction?
```

Comment:
This. Do we need to add a boolean to track whether we own the `BSP::Hitbox`?

Reply:
We will allow destruction to occur by the OS when program terminates. No need to manage destruction.